registry/etcdv3/service_discovery.go

```diff
@@ -282,7 +287,15 @@ func (e *etcdV3ServiceDiscovery) DataChange(eventType remoting.Event) bool {
 			instance.ServiceName = ""
 		}
 
-		if err := e.DispatchEventByServiceName(instance.ServiceName); err != nil {
+		// notify instance listener instance change
+		name := instance.ServiceName
+		instances := e.GetInstances(name)
+		for _, lis := range e.instanceListenerMap[instance.ServiceName].Values() {
```

Comment:
I am not sure if we should iterate all listeners even some of them return error.

Reply:
All listeners are of the same type need to be guaranteed. Or we need to change the definition of `instanceListenerMap map[string]*gxset.HashSet` in here, to avoid the defect of Golang does not support generic.

Reply:
`instanceListenerMap map[string]map[string][*registry.ServiceInstancesChangedListener]`

or if there's generic support

instanceListenerMap map[string]*gxset.HashSet<string, *registry.ServiceInstancesChangedListener>